README.md
# Rust ISL29125 Digital RGB Color Light Sensor with IR Blocking Filter Driver

[![crates.io](https://img.shields.io/crates/v/isl29125.svg)](https://crates.io/crates/isl29125)
[![Docs](https://docs.rs/isl29125/badge.svg)](https://docs.rs/isl29125)
[![Build Status](https://travis-ci.com/eldruin/isl29125-rs.svg?branch=master)](https://travis-ci.com/eldruin/isl29125-rs)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/isl29125-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/isl29125-rs?branch=master)

This is a platform agnostic Rust driver for the low-power digital RGB color
light sensor with IR blocking filter using the [`embedded-hal`] traits.

This driver allows you to:
- Read all colors. See: `read()`.
- Read red/green/blue colors individually. See: `red()`.
- Set operating mode. See: `set_operating_mode()`.
- Set ADC resolution. See: `set_resolution()`.
- Set RGB data sensing range. See: `set_range()`.
- Set IR filtering. See: `set_ir_filtering()`.
- Read the status flags. See: `status()`.
- Clear the status flags. See: `clear_status()`.
- Read the device ID. See: `device_id()`.
- Perform a software reset. See: `reset()`.
- Interrupts:
    - Set interrupt thresholds. See: `set_interrupt_thresholds()`.
    - Set interrupt threshold assignment. See: `set_interrupt_threshold_assignment()`.
    - Set the fault count. See: `set_fault_count()`.
    - Set interrupt pin mode. See: `set_interrupt_pin_mode()`.
    - Enable/Disable generating an interrupt after a conversion is done. See: `enable_interrupt_on_conversion_done()`.

<!-- TODO
[Introductory blog post]()
-->

## The device

The ISL29125 is a low power, high sensitivity, RED, GREEN and BLUE color
light sensor (RGB) with an I2C (SMBus compatible) interface. Its
state-of-the-art photodiode array provides an accurate RGB spectral
response and excellent light source to light source variation (LS2LS).

The ISL29125 is designed to reject IR in light sources allowing the device
to operate in environments from sunlight to dark rooms. The integrating
ADC rejects 50Hz and 60Hz flicker caused by artificial light sources.
A selectable range allows the user to optimize sensitivity suitable for
the specific application. In normal operation mode the device consumes
56μA, which reduces to 0.5μA in power-down mode.

The ISL29125 supports hardware and software user programmable interrupt
thresholds. The Interrupt persistency feature reduces false trigger
notification.

Datasheet: [ISL29125](https://www.renesas.com/eu/en/www/doc/datasheet/isl29125.pdf)


## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

Please find additional examples using hardware in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
extern crate linux_embedded_hal as hal;
use isl29125::{Isl29125, OperatingMode};

fn main() {
    let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Isl29125::new(dev);
    sensor
        .set_operating_mode(OperatingMode::RedGreenBlue)
        .unwrap();
    loop {
        let m = sensor.read().unwrap();
        println!("R: {}, G: {}, B: {}", m.red, m.green, m.blue);
    }
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/isl29125-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## 0.1.0 - 2020-04-14

Initial release of the functionally complete driver to crates.io.

[Unreleased]: https://github.com/eldruin/isl29125-rs/compare/v0.1.0...HEAD


tests/common/mod.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use isl29125::Isl29125;

pub struct Register;
#[allow(unused)]
impl Register {
    pub const DEVICE_ID: u8 = 0x00;
    pub const CONFIG1: u8 = 0x01;
    pub const CONFIG2: u8 = 0x02;
    pub const CONFIG3: u8 = 0x03;
    pub const THL: u8 = 0x04;
    pub const STATUS: u8 = 0x08;
    pub const GREEN_L: u8 = 0x09;
    pub const RED_L: u8 = 0x0B;
    pub const BLUE_L: u8 = 0x0D;
}

pub struct BitFlags;
impl BitFlags {
    pub const SYNC: u8 = 1 << 5;
    pub const RESOLUTION: u8 = 1 << 4;
    pub const RANGE: u8 = 1 << 3;
    pub const IR_OFFSET: u8 = 1 << 7;
    pub const CONVEN: u8 = 1 << 4;
    pub const BOUTF: u8 = 1 << 2;
    pub const CONVENF: u8 = 1 << 1;
    pub const RGBTHF: u8 = 1;
}

pub const ADDR: u8 = 0b100_0100;

#[allow(unused)]
pub fn new(transactions: &[I2cTrans]) -> Isl29125<I2cMock> {
    Isl29125::new(I2cMock::new(transactions))
}

#[allow(unused)]
pub fn destroy(sensor: Isl29125<I2cMock>) {
    sensor.destroy().done();
}


tests/integration.rs
mod common;
use crate::common::{destroy, new, BitFlags as BF, Register, ADDR};
use embedded_hal_mock::i2c::Transaction as I2cTrans;
use isl29125::{
    ConversionStatus as CS, FaultCount, IRFilteringRange, InterruptPinMode,
    InterruptThresholdAssignment, OperatingMode, Range, Resolution, Status,
};

#[test]
fn can_create_and_destroy() {
    let sensor = new(&[]);
    destroy(sensor);
}

#[test]
fn can_get_device_id() {
    let dev_id = 0xAB;
    let mut sensor = new(&[I2cTrans::write_read(
        ADDR,
        vec![Register::DEVICE_ID],
        vec![0xAB],
    )]);
    let id = sensor.device_id().unwrap();
    assert_eq!(dev_id, id);
    destroy(sensor);
}

#[test]
fn can_reset() {
    let mut sensor = new(&[I2cTrans::write(ADDR, vec![Register::DEVICE_ID, 0x46])]);
    sensor.reset().unwrap();
    destroy(sensor);
}

#[test]
fn can_get_measurement() {
    let red = 0x1234;
    let green = 0x5678;
    let blue = 0x9ABC;
    let mut sensor = new(&[I2cTrans::write_read(
        ADDR,
        vec![Register::GREEN_L],
        vec![0x78, 0x56, 0x34, 0x12, 0xBC, 0x9A],
    )]);
    let m = sensor.read().unwrap();
    assert_eq!(m.red, red);
    assert_eq!(m.green, green);
    assert_eq!(m.blue, blue);
    destroy(sensor);
}

macro_rules! set_test {
    ($name:ident, $method:ident, $reg:ident, $value:expr $(, $arg:expr)*) => {
        #[test]
        fn $name() {
            let mut sensor = new(&[I2cTrans::write(ADDR, vec![Register::$reg, $value])]);
            sensor.$method($($arg),*).unwrap();
            destroy(sensor);
        }
    };
}

macro_rules! set_operating_mode_test {
    ($name:ident, $value:expr, $mode:ident) => {
        set_test!(
            $name,
            set_operating_mode,
            CONFIG1,
            $value,
            OperatingMode::$mode
        );
    };
}
set_operating_mode_test!(set_mode_powerdown, 0, PowerDown);
set_operating_mode_test!(set_mode_green_only, 1, GreenOnly);
set_operating_mode_test!(set_mode_red_only, 2, RedOnly);
set_operating_mode_test!(set_mode_blue_only, 3, BlueOnly);
set_operating_mode_test!(set_mode_stand_by, 4, StandBy);
set_operating_mode_test!(set_mode_rgb, 5, RedGreenBlue);
set_operating_mode_test!(set_mode_red_green, 6, RedGreen);
set_operating_mode_test!(set_mode_green_blue, 7, GreenBlue);

set_test!(
    set_resolution_12,
    set_resolution,
    CONFIG1,
    BF::RESOLUTION,
    Resolution::Bit12
);
set_test!(
    set_resolution_16,
    set_resolution,
    CONFIG1,
    0,
    Resolution::Bit16
);

set_test!(set_range_375, set_range, CONFIG1, 0, Range::Lux375);
set_test!(
    set_range_10000,
    set_range,
    CONFIG1,
    BF::RANGE,
    Range::Lux10000
);

set_test!(
    set_int_mode_int,
    set_interrupt_pin_mode,
    CONFIG1,
    0,
    InterruptPinMode::Interrupt
);
set_test!(
    set_int_mode_sync,
    set_interrupt_pin_mode,
    CONFIG1,
    BF::SYNC,
    InterruptPinMode::SyncStart
);

#[test]
fn cannot_set_wrong_lower_ir_adjust() {
    let mut sensor = new(&[]);
    sensor
        .set_ir_filtering(IRFilteringRange::Lower(64))
        .expect_err("Should return error.");
    destroy(sensor);
}

#[test]
fn cannot_set_wrong_higher_ir_adjust() {
    let mut sensor = new(&[]);
    sensor
        .set_ir_filtering(IRFilteringRange::Higher(64))
        .expect_err("Should return error.");
    destroy(sensor);
}

set_test!(
    set_ir_filtering_lower,
    set_ir_filtering,
    CONFIG2,
    63,
    IRFilteringRange::Lower(63)
);

set_test!(
    set_ir_filtering_higher,
    set_ir_filtering,
    CONFIG2,
    BF::IR_OFFSET | 63,
    IRFilteringRange::Higher(63)
);

macro_rules! set_int_th_assign_test {
    ($name:ident, $value:expr, $mode:ident) => {
        set_test!(
            $name,
            set_interrupt_threshold_assignment,
            CONFIG3,
            $value,
            InterruptThresholdAssignment::$mode
        );
    };
}

set_int_th_assign_test!(set_int_th_assign_none, 0, None);
set_int_th_assign_test!(set_int_th_assign_green, 1, Green);
set_int_th_assign_test!(set_int_th_assign_red, 2, Red);
set_int_th_assign_test!(set_int_th_assign_blue, 3, Blue);

macro_rules! set_fc_test {
    ($name:ident, $value:expr, $mode:ident) => {
        set_test!($name, set_fault_count, CONFIG3, $value, FaultCount::$mode);
    };
}
set_fc_test!(set_fault_count_one, 0, One);
set_fc_test!(set_fault_count_two, 1 << 2, Two);
set_fc_test!(set_fault_count_four, 2 << 2, Four);
set_fc_test!(set_fault_count_eight, 3 << 2, Eight);

set_test!(
    enable_int_on_conv_done,
    enable_interrupt_on_conversion_done,
    CONFIG3,
    BF::CONVEN
);
set_test!(
    disable_int_on_conv_done,
    disable_interrupt_on_conversion_done,
    CONFIG3,
    0
);

#[test]
fn can_set_interrupt_thresholds() {
    let mut sensor = new(&[I2cTrans::write(
        ADDR,
        vec![Register::THL, 0x34, 0x12, 0x78, 0x56],
    )]);
    sensor.set_interrupt_thresholds(0x1234, 0x5678).unwrap();
    destroy(sensor);
}

macro_rules! get_status_test {
    ($name:ident, $value:expr, $expected_member:ident, $expected_value:expr) => {
        #[test]
        fn $name() {
            let mut sensor = new(&[I2cTrans::write_read(
                ADDR,
                vec![Register::STATUS],
                vec![$value],
            )]);
            let mut expected = Status::default();
            expected.$expected_member = $expected_value;
            let status = sensor.status().unwrap();
            assert_eq!(status, expected);
            destroy(sensor);
        }
    };
}

get_status_test!(get_status_default, 0, interrupt_triggered, false);
get_status_test!(get_status_int, BF::RGBTHF, interrupt_triggered, true);
get_status_test!(get_status_conv, BF::CONVENF, conversion_completed, true);
get_status_test!(get_status_brownout, BF::BOUTF, brownout, true);
get_status_test!(get_status_converting_g, 1 << 4, converting, CS::Green);
get_status_test!(get_status_converting_r, 2 << 4, converting, CS::Red);
get_status_test!(get_status_converting_b, 3 << 4, converting, CS::Blue);

set_test!(can_clear_status, clear_status, STATUS, 0);

macro_rules! get_color_test {
    ($name:ident, $register:ident, $method:ident) => {
        #[test]
        fn $name() {
            let mut sensor = new(&[I2cTrans::write_read(
                ADDR,
                vec![Register::$register],
                vec![0x34, 0x12],
            )]);
            let value = sensor.$method().unwrap();
            assert_eq!(value, 0x1234);
            destroy(sensor);
        }
    };
}
get_color_test!(get_color_red, RED_L, red);
get_color_test!(get_color_green, GREEN_L, green);
get_color_test!(get_color_blue, BLUE_L, blue);


src/lib.rs
//! This is a platform agnostic Rust driver for the low-power digital RGB color
//! light sensor with IR blocking filter using the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Read all colors. See: [`read()`].
//! - Read red/green/blue colors individually. See: [`red()`].
//! - Set operating mode. See: [`set_operating_mode()`].
//! - Set ADC resolution. See: [`set_resolution()`].
//! - Set RGB data sensing range. See: [`set_range()`].
//! - Set IR filtering. See: [`set_ir_filtering()`].
//! - Read the status flags. See: [`status()`].
//! - Clear the status flags. See: [`clear_status()`].
//! - Read the device ID. See: [`device_id()`].
//! - Perform a software reset. See: [`reset()`].
//! - Interrupts:
//!     - Set interrupt thresholds. See: [`set_interrupt_thresholds()`].
//!     - Set interrupt threshold assignment. See: [`set_interrupt_threshold_assignment()`].
//!     - Set the fault count. See: [`set_fault_count()`].
//!     - Set interrupt pin mode. See: [`set_interrupt_pin_mode()`].
//!     - Enable/Disable generating an interrupt after a conversion is done. See: [`enable_interrupt_on_conversion_done()`].
//!
//! [`read()`]: struct.Isl29125.html#method.read
//! [`red()`]: struct.Isl29125.html#method.red
//! [`set_operating_mode()`]: struct.Isl29125.html#method.set_operating_mode
//! [`set_resolution()`]: struct.Isl29125.html#method.set_resolution
//! [`set_range()`]: struct.Isl29125.html#method.set_range
//! [`set_ir_filtering()`]: struct.Isl29125.html#method.set_ir_filtering
//! [`status()`]: struct.Isl29125.html#method.status
//! [`clear_status()`]: struct.Isl29125.html#method.clear_status
//! [`device_id()`]: struct.Isl29125.html#method.device_id
//! [`reset()`]: struct.Isl29125.html#method.reset
//! [`set_interrupt_thresholds()`]: struct.Isl29125.html#method.set_interrupt_thresholds
//! [`set_interrupt_threshold_assignment()`]: struct.Isl29125.html#method.set_interrupt_threshold_assignment
//! [`set_fault_count()`]: struct.Isl29125.html#method.set_fault_count
//! [`set_interrupt_pin_mode()`]: struct.Isl29125.html#method.set_interrupt_pin_mode
//! [`enable_interrupt_on_conversion_done()`]: struct.Isl29125.html#method.enable_interrupt_on_conversion_done
//!
//!
//! <!-- TODO
//! [Introductory blog post](TODO)
//! -->
//!
//! ## The device
//!
//! The ISL29125 is a low power, high sensitivity, RED, GREEN and BLUE color
//! light sensor (RGB) with an I2C (SMBus compatible) interface. Its
//! state-of-the-art photodiode array provides an accurate RGB spectral
//! response and excellent light source to light source variation (LS2LS).
//!
//! The ISL29125 is designed to reject IR in light sources allowing the device
//! to operate in environments from sunlight to dark rooms. The integrating
//! ADC rejects 50Hz and 60Hz flicker caused by artificial light sources.
//! A selectable range allows the user to optimize sensitivity suitable for
//! the specific application. In normal operation mode the device consumes
//! 56μA, which reduces to 0.5μA in power-down mode.
//!
//! The ISL29125 supports hardware and software user programmable interrupt
//! thresholds. The Interrupt persistency feature reduces false trigger
//! notification.
//!
//! Datasheet: [ISL29125](https://www.renesas.com/eu/en/www/doc/datasheet/isl29125.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the device.
//!
//! Please find additional examples using hardware in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Enable reading RGB and print the values
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! use isl29125::{Isl29125, OperatingMode};
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Isl29125::new(dev);
//! sensor
//!     .set_operating_mode(OperatingMode::RedGreenBlue)
//!     .unwrap();
//! loop {
//!     let m = sensor.read().unwrap();
//!     println!("R: {}, G: {}, B: {}", m.red, m.green, m.blue);
//! }
//! # }
//! ```
//!
//! ### Measure only the red color
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! use isl29125::{Isl29125, OperatingMode};
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Isl29125::new(dev);
//! sensor
//!     .set_operating_mode(OperatingMode::RedOnly)
//!     .unwrap();
//! loop {
//!     let red = sensor.red().unwrap();
//!     println!("Red: {}", red);
//! }
//! # }
//! ```
//!
//! ### Print the green color when the thresholds are exceeded 4 times
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! use isl29125::{
//!     Isl29125, OperatingMode, FaultCount, InterruptThresholdAssignment
//! };
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Isl29125::new(dev);
//! sensor
//!     .set_operating_mode(OperatingMode::GreenOnly)
//!     .unwrap();
//! let ita = InterruptThresholdAssignment::Green;
//! sensor.set_interrupt_threshold_assignment(ita).unwrap();
//! sensor.set_fault_count(FaultCount::Four).unwrap();
//! sensor.set_interrupt_thresholds(150, 8500).unwrap();
//! loop {
//!     while !sensor.status().unwrap().interrupt_triggered {}
//!     let green = sensor.green().unwrap();
//!     println!("Green: {}", green);
//! }
//! # }
//! ```
//!
//! ### Set the IR filtering
//!
//! ```no_run
//! extern crate linux_embedded_hal as hal;
//! use isl29125::{Isl29125, IRFilteringRange};
//!
//! # fn main() {
//! let dev = hal::I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Isl29125::new(dev);
//! sensor
//!     .set_ir_filtering(IRFilteringRange::Lower(35))
//!     .unwrap();
//! # }
//! ```
//!
#![deny(unsafe_code, missing_docs)]
#![no_std]

mod device_impl;
mod types;
pub use crate::types::{
    ConversionStatus, Error, FaultCount, IRFilteringRange, InterruptPinMode,
    InterruptThresholdAssignment, Measurement, OperatingMode, Range, Resolution, Status,
};
mod register_address;
use crate::register_address::{BitFlags, Register};

/// ISL29125 device driver
#[derive(Debug)]
pub struct Isl29125<I2C> {
    i2c: I2C,
    config1: Config,
    config3: Config,
}

#[derive(Debug, Default, Clone, Copy)]
struct Config {
    bits: u8,
}


src/types.rs
/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C communication error
    I2C(E),
    /// Invalid input data provided
    InvalidInputData,
}

/// Measurement result
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Measurement {
    /// Red
    pub red: u16,
    /// Green
    pub green: u16,
    /// Blue
    pub blue: u16,
}

/// Operating mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum OperatingMode {
    /// Power-Down (ADC conversion)
    PowerDown,
    /// Stand by (No ADC conversion)
    StandBy,
    /// Red only
    RedOnly,
    /// Green only
    GreenOnly,
    /// Blue only
    BlueOnly,
    /// Red/Green
    RedGreen,
    /// Green/Blue
    GreenBlue,
    /// Red/Green/Blue
    RedGreenBlue,
}

impl Default for OperatingMode {
    fn default() -> Self {
        OperatingMode::PowerDown
    }
}

/// ADC resolution
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Resolution {
    /// 12-bit resolution
    Bit12,
    /// 16-bit resolution (default)
    Bit16,
}

impl Default for Resolution {
    fn default() -> Self {
        Resolution::Bit16
    }
}

/// RGB data sensing range
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Range {
    /// 375 lux
    Lux375,
    /// 10000 lux
    Lux10000,
}

impl Default for Range {
    fn default() -> Self {
        Range::Lux375
    }
}

/// Interrupt pin (INT) mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum InterruptPinMode {
    /// Interrupts will be generated on exceeded thresholds. (ADC starts when writing to the config 1 register)
    Interrupt,
    /// INT pin is an input. ADC conversion starts on the rising edge at the INT pin.
    SyncStart,
}

impl Default for InterruptPinMode {
    fn default() -> Self {
        InterruptPinMode::Interrupt
    }
}

/// IR filtering range
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IRFilteringRange {
    /// Lower range
    ///
    /// The value must be in the range `[0-63]` (no offset)
    Lower(u8),
    /// Higher range
    ///
    /// The value must be in the range `[0-63]` (this corresponds to 106-169)
    Higher(u8),
}

impl Default for IRFilteringRange {
    fn default() -> Self {
        IRFilteringRange::Lower(0)
    }
}

/// Interrupt threshold assignment
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum InterruptThresholdAssignment {
    /// No interrupt (default)
    None,
    /// Use interrupt thresholds on red channel data.
    Red,
    /// Use interrupt thresholds on green channel data.
    Green,
    /// Use interrupt thresholds on blue channel data.
    Blue,
}

impl Default for InterruptThresholdAssignment {
    fn default() -> Self {
        InterruptThresholdAssignment::None
    }
}

/// Fault count
///
/// Number of consecutive fault events necessary to trigger interrupt.
/// This is referred to as "persistence" in the documentation.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum FaultCount {
    /// One (default)
    One,
    /// Two
    Two,
    /// Four
    Four,
    /// Eight
    Eight,
}

impl Default for FaultCount {
    fn default() -> Self {
        FaultCount::One
    }
}

/// Status
#[derive(Debug, Default, Clone, Copy, PartialEq)]
pub struct Status {
    /// Whether an interrupt was triggered
    pub interrupt_triggered: bool,
    /// Whether a conversion was completed
    pub conversion_completed: bool,
    /// Whether a power-down or brownout condition occurred
    pub brownout: bool,
    /// Wheter a color channel is under conversion
    pub converting: ConversionStatus,
}

/// RGB conversion status
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ConversionStatus {
    /// No operation
    NoOperation,
    /// Red
    Red,
    /// Green
    Green,
    /// Blue
    Blue,
}

impl Default for ConversionStatus {
    fn default() -> Self {
        ConversionStatus::NoOperation
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_get_default_operating_mode() {
        assert_eq!(OperatingMode::PowerDown, OperatingMode::default());
    }

    #[test]
    fn can_get_default_resolution() {
        assert_eq!(Resolution::Bit16, Resolution::default());
    }

    #[test]
    fn can_get_default_range() {
        assert_eq!(Range::Lux375, Range::default());
    }

    #[test]
    fn can_get_default_int_pin_mode() {
        assert_eq!(InterruptPinMode::Interrupt, InterruptPinMode::default());
    }

    #[test]
    fn can_get_default_ir_filtering_range() {
        assert_eq!(IRFilteringRange::Lower(0), IRFilteringRange::default());
    }

    #[test]
    fn can_get_default_int_threshold_assignment() {
        assert_eq!(
            InterruptThresholdAssignment::None,
            InterruptThresholdAssignment::default()
        );
    }

    #[test]
    fn can_get_default_fault_count() {
        assert_eq!(FaultCount::One, FaultCount::default());
    }

    #[test]
    fn can_get_default_status() {
        assert_eq!(
            Status {
                interrupt_triggered: false,
                conversion_completed: false,
                brownout: false,
                converting: ConversionStatus::NoOperation
            },
            Status::default()
        );
    }
}


src/register_address.rs
use crate::{Error, Isl29125};
use embedded_hal::blocking::i2c;

pub const ADDR: u8 = 0b100_0100;

pub struct Register;
impl Register {
    pub const DEVICE_ID: u8 = 0x00;
    pub const CONFIG1: u8 = 0x01;
    pub const CONFIG2: u8 = 0x02;
    pub const CONFIG3: u8 = 0x03;
    pub const THL: u8 = 0x04;
    pub const STATUS: u8 = 0x08;
    pub const GREEN_L: u8 = 0x09;
    pub const RED_L: u8 = 0x0B;
    pub const BLUE_L: u8 = 0x0D;
}

pub struct BitFlags;
impl BitFlags {
    pub const SYNC: u8 = 1 << 5;
    pub const RESOLUTION: u8 = 1 << 4;
    pub const RANGE: u8 = 1 << 3;
    pub const IR_OFFSET: u8 = 1 << 7;
    pub const CONVEN: u8 = 1 << 4;
    pub const BOUTF: u8 = 1 << 2;
    pub const CONVENF: u8 = 1 << 1;
    pub const RGBTHF: u8 = 1;
}

impl<I2C, E> Isl29125<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    pub(crate) fn write_register(&mut self, register: u8, data: u8) -> Result<(), Error<E>> {
        let payload: [u8; 2] = [register, data];
        self.i2c.write(ADDR, &payload).map_err(Error::I2C)
    }

    pub(crate) fn write_thresholds(&mut self, low: u16, high: u16) -> Result<(), Error<E>> {
        let payload: [u8; 5] = [
            Register::THL,
            (low & 0xFF) as u8,
            ((low & 0xFF00) >> 8) as u8,
            (high & 0xFF) as u8,
            ((high & 0xFF00) >> 8) as u8,
        ];
        self.i2c.write(ADDR, &payload).map_err(Error::I2C)
    }
}

impl<I2C, E> Isl29125<I2C>
where
    I2C: i2c::WriteRead<Error = E>,
{
    pub(crate) fn read_register(&mut self, register: u8) -> Result<u8, Error<E>> {
        let mut data = [0];
        self.read_data(register, &mut data).and(Ok(data[0]))
    }

    pub(crate) fn read_color(&mut self, register: u8) -> Result<u16, Error<E>> {
        let mut data = [0; 2];
        self.read_data(register, &mut data)
            .and(Ok(u16::from(data[0]) | (u16::from(data[1]) << 8)))
    }

    pub(crate) fn read_data(&mut self, register: u8, data: &mut [u8]) -> Result<(), Error<E>> {
        self.i2c
            .write_read(ADDR, &[register], data)
            .map_err(Error::I2C)
    }
}
